cassovary-core/src/main/scala/com/twitter/cassovary/util/collections/FastMap.scala

```diff
@@ -0,0 +1,428 @@
+package com.twitter.cassovary.util.collections
+
+import java.util
+
+import it.unimi.dsi.fastutil.ints._
+import it.unimi.dsi.fastutil.objects._
+
+import scala.collection.generic.CanBuildFrom
+import scala.collection.mutable
+import scala.{specialized => sp}
+
+/**
+ * Specialized, fast and memory efficient implementation of mutable map.
+ *
+ * Provides basic operations only. See [[FastMapWrapper]] for how to use
+ * it as a [[mutable.Map]].
+ *
+ * Use constructor provided in the companion object.
+ */
+trait FastMap[@sp(Int, Long, AnyRef) A, @sp(Int, Long, AnyRef) B] {
+
+  /**
+   * Adds key-value pair to the map.
+   */
+  def +=(key: A, value: B): this.type
+
+  /**
+   * Removes key-value pair with a given key.
+   */
+  def -=(key: A): this.type
+
+  /**
+   * Get is optimized for retrieving elements contained in the Map.
+   *
+   * If the element is not in the map, returns `0` of given type (or false, null).
+   */
+  def get(key: A): B
+
+  def clear(): Unit
+
+  def contains(key: A): Boolean
+
+  def size(): Int
+
+  /**
+   * Wraps as a [[mutable.Map]].
+   */
+  def asScala(): mutable.Map[A, B] = new FastMapWrapper[A, B](this)
+
+  /**
+   * Returns iterator over key, value pairs stored in the map.
+   *
+   * Used only inside to conform to Scala's collections.
+   */
+  private[collections] def iterator: Iterator[(A, B)]
+}
+
+/**
+ * Implements incrementing value specified by a given key.
+ */
+trait AddTo[@sp(Int, Long, AnyRef) A] {
+  def addTo(key: A, increment: Int)
+}
+
+/**
+ * Marks a [[FastMap]] with an iterator that preserves insertion order.
+ */
+trait InsertionOrderIterator
+
+/**
+ * Features creation of specialized FastMaps.
+ */
+object FastMap {
+  /**
+   * @return Specialized FastMap created for key type `A` and value type `B`.
+   */
+  def apply[@sp(Int, Long, AnyRef) A, @sp(Int, Long, AnyRef) B]()
+      (implicit impl: FastMapFactory[A, B, FastMap[A, B]]): FastMap[A, B] = {
+    impl()
+  }
+
+  /**
+   * @return FastMap created for given key type `A`, value type `B` and FastMap type `MapType`.
+   */
+  def applyFor[@sp(Int, Long, AnyRef) A, @sp(Int, Long, AnyRef) B, MapType <: FastMap[A, B]]()
```

Comment:
In a simple case I create map using just: `FastMap[Int, Int]()`
But then we have a problem, when we need addTo or preserved insertion order... Then we need to specify also the returned map type and we have `FastMap.applyFor[Int, Int, Map[Int, Int] with AddTo]()